Source/WebCore/platform/ios/WebEvent.mm

```diff
@@ -213,6 +218,72 @@ - (WebEvent *)initWithKeyEventType:(WebEventType)type
     return self;
 }
 
+#if HAVE(UI_ASYNC_TEXT_INPUT)
+
+static inline WebEventType webEventType(UIKeyEventType type)
+{
+    switch (type) {
+    case UIKeyEventKeyDown:
+        return WebEventKeyDown;
+    case UIKeyEventKeyUp:
+        return WebEventKeyUp;
+    default:
+        ASSERT_NOT_REACHED();
+        return WebEventKeyDown;
+    }
+}
+
+static inline WebEventFlags webEventModifierFlags(UIKeyModifierFlags flags)
+{
+    WebEventFlags modifiers = 0;
+    if (flags & UIKeyModifierCommand)
+        modifiers |= WebEventFlagMaskCommandKey;
+    if (flags & UIKeyModifierAlternate)
+        modifiers |= WebEventFlagMaskOptionKey;
+    if (flags & UIKeyModifierControl)
+        modifiers |= WebEventFlagMaskControlKey;
+    if (flags & UIKeyModifierShift)
+        modifiers |= WebEventFlagMaskShiftKey;
+    return modifiers;
+}
+
+- (WebEvent *)initWithUIKeyEvent:(UIKeyEvent *)event
+{
+    if (!(self = [super init]))
+        return nil;
+
+    _type = webEventType(event.type);
+    _timestamp = static_cast<CFTimeInterval>(event.timestamp);
+    _modifierFlags = webEventModifierFlags(event.modifierFlags);
+    _keyboardFlags = 0;
+    // FIXME: Set WebEventKeyboardInputModifierFlagsChanged as needed.
+    if (event.keyRepeating)
+        _keyboardFlags |= WebEventKeyboardInputRepeat;
+
+    auto keyCode = static_cast<uint16_t>(event.keyCode);
+    auto characters = event.characters;
+    auto charactersIgnoringModifiers = event.charactersIgnoringModifiers;
+    if (keyCode)
+        _keyCode = windowsKeyCodeForKeyCode(keyCode);
+    else if (charactersIgnoringModifiers.length == 1)
+        _keyCode = windowsKeyCodeForCharCodeIOS([charactersIgnoringModifiers characterAtIndex:0]);
+
+    _characters = [normalizedStringWithAppKitCompatibilityMapping(characters, keyCode) retain];
```

Comment:
manual retains! weird

Reply:
...yeah :(

I _really_ wanted to try and turn `_characters`, `_charactersIgnoringModifiers` into `RetainPtr<NSString>`, but that would break AX I think, which currently relies on these `@private` ivars.